docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,216 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+
+### Benefits of Using Pre-Shared Keys
+
+- Using pre-shared keys can avoid the need for public key operations.
+  This is useful in performance-constrained environments with limited CPU power.
+- PSKs may be more convenient from a key management point of view. For instance, in closed environments where the connections
+  are configured manually in advance, it is easier to configure a PSK than to use certificates. Another case is when the parties already have a mechanism for setting up a shared secret key, and that mechanism could be used to "bootstrap" a key for authenticating the TLS connection.
+- PSKs are more appropriate for use cases with long-lived credentials stored in hardware secure elements than PKI-based
+  solutions which enforces expiration dates and requires error-prone rotation schemes.
+
+### Security Considerations
+
+- To prevent the reflection/selfie attack use a different PSK for each endpoint, and use a different PSK when acting as a server and a different PSK when acting as a client. Read the research paper titled [Selfie: reflections on TLS 1.3 with PSK](https://eprint.iacr.org/2019/347.pdf) for more information.
+- The PSK should be set with the server name in mind. If the server name is configured during the ClientHello callback, then so should the PSKs.
+- The external PSK authentication mechanism implicitly assumes each PSK is known to exactly one client and one server, and these peers never switch roles. If this assumption is violated, the security properties of >= TLS1.3 are severely weakened.
+
+### Configuring External Pre-Shared Keys
+
+Use the following APIs to configure and use external pre-shared keys.
+
+```c
+struct s2n_psk* s2n_external_psk_new();
+int s2n_psk_free(struct s2n_psk **psk);
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size);
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac);
+int s2n_connection_append_psk(struct s2n_connection *conn, struct s2n_psk *psk);
+int s2n_config_set_psk_mode(struct s2n_config *config, s2n_psk_mode mode);
+int s2n_connection_set_psk_mode(struct s2n_connection *conn, s2n_psk_mode mode);
+```
+
+To create an external PSK a caller must call the **s2n_external_psk_new** API. **s2n_external_psk_new** creates a new external PSK object
+with **S2N_PSK_HMAC_SHA256** as the default PSK hash algorithm. Use **s2n_psk_free** to free the memory allocated to the external
+PSK object created by this API.
+
+**s2n_psk_set_identity** sets the PSK identity for a given PSK. The PSK identity is a unique identifier for the pre-shared secret. The PSK identity is transmitted over the network unencrypted and is a non-secret value. Do not include any confidential information in the identity. Note that the identity value is copied into s2n-tls memory and the user is responsible for freeing the memory associated with the **identity** input.
```

Comment:
```suggestion
**s2n_psk_set_identity** sets the PSK identity for a given PSK. The PSK identity is a unique identifier for the pre-shared secret. This identity is transmitted over the network unencrypted and is a non-secret value so do not include any confidential information in the identity. 
```
I thought we were putting memory safety information in the doxygen comments, but 🤷‍♀️ I'm flexible here.

Reply:
The Usage Guide is meant to help the customer with gotchas and things they should be aware of while using the API, information such as memory responsibility  fits here for that reason.